Source/Diagnostics/ReducedDiags/ParticleMomentum.cpp

```diff
@@ -0,0 +1,262 @@
+            ofs << m_sep;
+            ofs << "[" << c++ << "]total_z(kg*m/s)";
+
+            for (int i = 0; i < nSpecies; ++i)
+            {
+                ofs << m_sep;
+                ofs << "[" << c++ << "]";
+                ofs << species_names[i] + "_x(kg*m/s)";
+                ofs << m_sep;
+                ofs << "[" << c++ << "]";
+                ofs << species_names[i] + "_y(kg*m/s)";
+                ofs << m_sep;
+                ofs << "[" << c++ << "]";
+                ofs << species_names[i] + "_z(kg*m/s)";
+            }
+
+            ofs << m_sep;
+            ofs << "[" << c++ << "]";
+            ofs << "total_mean_x(kg*m/s)";
+            ofs << m_sep;
+            ofs << "[" << c++ << "]";
+            ofs << "total_mean_y(kg*m/s)";
+            ofs << m_sep;
+            ofs << "[" << c++ << "]";
+            ofs << "total_mean_z(kg*m/s)";
+
+            for (int i = 0; i < nSpecies; ++i)
+            {
+                ofs << m_sep;
+                ofs << "[" << c++ << "]";
+                ofs << species_names[i] + "_mean_x(kg*m/s)";
+                ofs << m_sep;
+                ofs << "[" << c++ << "]";
+                ofs << species_names[i] + "_mean_y(kg*m/s)";
+                ofs << m_sep;
+                ofs << "[" << c++ << "]";
+                ofs << species_names[i] + "_mean_z(kg*m/s)";
+            }
+
+            ofs << std::endl;
+            ofs.close();
+        }
+    }
+}
+
+void ParticleMomentum::ComputeDiags (int step)
+{
+    // Check if the diags should be done
+    if (m_intervals.contains(step+1) == false)
+    {
+        return;
+    }
+
+    // Get MultiParticleContainer class object
+    const auto & mypc = WarpX::GetInstance().GetPartContainer();
+
+    // Get number of species
+    const int nSpecies = mypc.nSpecies();
+
+    // Get species names
+    const std::vector<std::string> species_names = mypc.GetSpeciesNames();
+
+    // Some useful offsets to fill m_data below
+    int offset_total_species, offset_mean_species, offset_mean_all;
+
+    // Loop over species
+    for (int i_s = 0; i_s < nSpecies; ++i_s)
+    {
+        // Get WarpXParticleContainer class object
+        const auto & myspc = mypc.GetParticleContainer(i_s);
+
+        // Get mass
+        // (photons must be treated as a special case: they have zero mass,
+        // but ux, uy, uz are calculated assuming a mass equal to the electron mass)
+        const amrex::Real m = (myspc.AmIA<PhysicalSpecies::photon>()) ? PhysConst::m_e : myspc.getMass();
+
+        using PType = typename WarpXParticleContainer::SuperParticleType;
+
+        // Use amrex::ReduceSum to compute the sum of the momentum of all particles held by
+        // the current MPI rank for this species (loop over all boxes held by this MPI rank)
+        amrex::Real Px = 0.0_rt;
+        amrex::Real Py = 0.0_rt;
+        amrex::Real Pz = 0.0_rt;
+
+        Px = ReduceSum(myspc, [=] AMREX_GPU_HOST_DEVICE (const PType& p) -> Real
```

Comment:
I think we can pass multiple lambdas here, but am not sure if this creates only one kernel.

I would avoid iterating the particle species multiple times, because that is generally a costly operation. Can you try to
create a single lambda that returns `Real[4]` for the three components and sum of `w`?

Reply:
Yes, thank you for the suggestion. I'll mark the PR as "[WIP]" again and try to see if I can do that.

Reply:
@ax3l As discussed on Slack, this cannot be done currently, but @atmyers plans to add support for it in the future.

Reply:
I used the new tuple-level reductions for particle containers in https://github.com/ECP-WarpX/WarpX/pull/1908/commits/fbec288743f2bcc5b5edb2d65c3bb780d9dac86b, so now we have a single lambda that returns all necessary quantities.